File: examples/tlog_record.rs
use mavkit::Vehicle;
use mavkit::tlog::TlogWriter;
use mavlink::MavlinkVersion;
use std::io::BufWriter;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let bind_addr = std::env::args()
        .nth(1)
        .unwrap_or_else(|| "0.0.0.0:14550".to_string());
    let path = std::env::args()
        .nth(2)
        .unwrap_or_else(|| "recording.tlog".to_string());

    println!("connecting to {bind_addr}...");
    let vehicle = Vehicle::connect_udp(&bind_addr).await?;
    println!("connected, recording to {path}");

    let mut rx = vehicle.raw_messages();
    let file = std::fs::File::create(&path)?;
    let mut writer = TlogWriter::new(BufWriter::new(file), MavlinkVersion::V2);

    let mut count = 0u64;
    loop {
        match rx.recv().await {
            Ok((header, msg)) => {
                writer.write_now(&header, &msg)?;
                count += 1;
                if count.is_multiple_of(100) {
                    writer.flush()?;
                    println!("{count} messages recorded");
                }
            }
            Err(tokio::sync::broadcast::error::RecvError::Lagged(n)) => {
                eprintln!("warning: skipped {n} messages (receiver lagged)");
            }
            Err(tokio::sync::broadcast::error::RecvError::Closed) => {
                break;
            }
        }
    }

    writer.flush()?;
    println!("{count} messages total");
    Ok(())
}

File: mavkit-python/examples/tlog_record.py
"""Record raw MAVLink messages to a TLOG file."""

import asyncio
import os
import sys

import mavkit


async def main():
    bind_addr = os.environ.get("MAVKIT_EXAMPLE_UDP_BIND", "0.0.0.0:14550")
    path = sys.argv[1] if len(sys.argv) > 1 else "recording.tlog"

    vehicle = await mavkit.Vehicle.connect_udp(bind_addr)
    try:
        raw = vehicle.raw()
        stream = raw.subscribe()

        with mavkit.TlogWriter(path) as writer:
            print(f"connected, recording to {path}")
            count = 0
            async for message in stream:
                _ = writer.write(message)
                count += 1
                if count % 100 == 0:
                    writer.flush()
                    print(f"{count} messages recorded")
    finally:
        await vehicle.disconnect()


asyncio.run(main())
